extra/modules/greenbids-real-time-data/src/main/java/org/prebid/server/hooks/modules/greenbids/real/time/data/v1/GreenbidsRealTimeDataProcessedAuctionRequestHook.java

```diff
@@ -0,0 +1,195 @@
+            AuctionRequestPayload auctionRequestPayload,
+            AuctionInvocationContext invocationContext) {
+
+        final AuctionContext auctionContext = invocationContext.auctionContext();
+        final BidRequest bidRequest = auctionContext.getBidRequest();
+        final Partner partner = parseBidRequestExt(bidRequest);
+
+        if (partner == null) {
+            return Future.succeededFuture(toInvocationResult(
+                    bidRequest, null, InvocationAction.no_action));
+        }
+
+        return onnxModelRunnerWithThresholds.retrieveOnnxModelRunner(partner)
+                .compose(onnxModelRunner -> onnxModelRunnerWithThresholds.retrieveThreshold(partner, mapper)
+                        .compose(throttlingThresholds -> {
+                            Map<String, Map<String, Boolean>> impsBiddersFilterMap = null;
+                            try {
+                                if (onnxModelRunner == null || throttlingThresholds == null) {
+                                    throw new PreBidException(
+                                            "Cache was empty, fetching and put artefacts for next request");
+                                }
+
+                                final GreenbidsInferenceData greenbidsInferenceData = GreenbidsInferenceData
+                                        .prepareData(bidRequest, database, mapper);
+
+                                final Double threshold = partner.getThreshold(throttlingThresholds);
+                                impsBiddersFilterMap = filterService.runModeAndFilterBidders(
+                                        onnxModelRunner,
+                                        greenbidsInferenceData.getThrottlingMessages(),
+                                        greenbidsInferenceData.getThrottlingInferenceRows(),
+                                        threshold);
+                            } catch (PreBidException e) {
+                                return Future.succeededFuture(toInvocationResult(
+                                        bidRequest, null, InvocationAction.no_action));
+                            }
+
+                            final GreenbidsInvocationResult greenbidsInvocationResult = GreenbidsInvocationResult
+                                    .prepareInvocationResult(
+                                            partner, bidRequest, impsBiddersFilterMap);
+
+                            return Future.succeededFuture(toInvocationResult(
+                                    greenbidsInvocationResult.getUpdatedBidRequest(),
+                                    greenbidsInvocationResult.getAnalyticsResult(),
+                                    greenbidsInvocationResult.getInvocationAction()));
+                        }))
+                .recover(throwable -> Future.succeededFuture(toInvocationResult(
+                        bidRequest, null, InvocationAction.no_action)));
+    }
```

Comment:
```java
    @Override
    public Future<InvocationResult<AuctionRequestPayload>> call(
            AuctionRequestPayload auctionRequestPayload,
            AuctionInvocationContext invocationContext) {

        final AuctionContext auctionContext = invocationContext.auctionContext();
        final BidRequest bidRequest = auctionContext.getBidRequest();
        final Partner partner = parseBidRequestExt(bidRequest);

        if (partner == null) {
            return Future.succeededFuture(toInvocationResult(
                    bidRequest, null, InvocationAction.no_action));
        }

        return Future.join(
                        onnxModelRunnerWithThresholds.retrieveOnnxModelRunner(partner),
                        onnxModelRunnerWithThresholds.retrieveThreshold(partner, mapper))
                .compose(compositeFuture -> toInvocationResult(
                        bidRequest,
                        partner,
                        compositeFuture.resultAt(0),
                        compositeFuture.resultAt(1)))
                .recover(throwable -> Future.succeededFuture(
                        toInvocationResult(bidRequest, null, InvocationAction.no_action)));
    }

    private Future<InvocationResult<AuctionRequestPayload>> toInvocationResult(BidRequest bidRequest,
                                                                               Partner partner,
                                                                               OnnxModelRunner onnxModelRunner,
                                                                               Double throttlingThreshold) {

        final Map<String, Map<String, Boolean>> impsBiddersFilterMap;
        try {
            final GreenbidsInferenceData greenbidsInferenceData = GreenbidsInferenceData.of(
                    bidRequest,
                    database,
                    mapper);

            impsBiddersFilterMap = filterService.runModeAndFilterBidders(
                    onnxModelRunner,
                    greenbidsInferenceData.getThrottlingMessages(),
                    greenbidsInferenceData.getThrottlingInferenceRows(),
                    throttlingThreshold);
        } catch (PreBidException e) {
            return Future.succeededFuture(toInvocationResult(bidRequest, null, InvocationAction.no_action));
        }

        final GreenbidsInvocationResult greenbidsInvocationResult = GreenbidsInvocationResult.of(
                partner,
                bidRequest,
                impsBiddersFilterMap);

        return Future.succeededFuture(toInvocationResult(
                greenbidsInvocationResult.getUpdatedBidRequest(),
                greenbidsInvocationResult.getAnalyticsResult(),
                greenbidsInvocationResult.getInvocationAction()));
    }

    private InvocationResult<AuctionRequestPayload> toInvocationResult(
            BidRequest bidRequest,
            AnalyticsResult analyticsResult,
            InvocationAction action) {

        final List<AnalyticsResult> analyticsResults = analyticsResult == null
                ? Collections.emptyList()
                : Collections.singletonList(analyticsResult);

        return InvocationResultImpl
                .<AuctionRequestPayload>builder()
                .status(InvocationStatus.success)
                .action(action)
                .payloadUpdate(payload -> AuctionRequestPayloadImpl.of(bidRequest))
                .analyticsTags(toAnalyticsTags(analyticsResults))
                .build();
    }
```

Reply:
Refactored using `Future.join` 

Also added annotation `Value(staticConstructor = "of")` and static method `of` to `GreenbidsInferenceData ` and `GreenbidsInvocationResult`